Lib/pdb.py

```diff
@@ -2491,6 +2499,490 @@ def set_trace(*, header=None, commands=None):
+                }
+            }:
+                self.bpnum = bpnum
+                self.command_list_terminators = command_list_terminators
+                self.commands_mode = True
+            case _:
+                raise RuntimeError(f"Unrecognized payload {payload}")
+
+    def prompt_for_breakpoint_command_list(self, prompt):
+        parts = [f"commands {self.bpnum}"]
+        while True:
+            try:
+                line = input(prompt).strip()
+                parts.append(line)
+                cmd = self.pdb_instance.parseline(line)[0]
+                if cmd in self.command_list_terminators:
+                    break
+            except EOFError:
+                return
+            finally:
+                self.commands_mode = False
+
+        command = "\n".join(parts)
+        self.sockfile.write((json.dumps({"command": command}) + "\n").encode())
+        self.sockfile.flush()
+
+    def prompt_for_repl_command(self, prompt):
+        while True:
+            try:
+                command = self.read_command(prompt)
+            except EOFError:
+                command = "EOF"
+            except KeyboardInterrupt:
+                print(flush=True)
+                continue
+            except Exception as exc:
+                msg = traceback.format_exception_only(exc)[-1].strip()
+                print("***", msg, flush=True)
+                continue
+
+            if self.interact_mode:
+                payload = {"interact": command}
+            else:
+                payload = {"command": command}
+            self.sockfile.write((json.dumps(payload) + "\n").encode())
+            self.sockfile.flush()
+            return
+
+    def complete(self, text, state):
+        import readline
+
+        if state == 0:
+            origline = readline.get_line_buffer()
+            line = origline.lstrip()
+            stripped = len(origline) - len(line)
+            begidx = readline.get_begidx() - stripped
+            endidx = readline.get_endidx() - stripped
+
+            msg = {
+                "completion": {
+                    "text": text,
+                    "line": line,
+                    "begidx": begidx,
+                    "endidx": endidx,
+                }
+            }
+            self.sockfile.write((json.dumps(msg) + "\n").encode())
+            self.sockfile.flush()
+
+            payload = self.sockfile.readline()
+            payload = json.loads(payload)
+            if "completion" not in payload:
+                raise RuntimeError(
+                    f"Failed to get valid completions. Got: {payload}"
+                )
+
+            self.completion_matches = payload["completion"]
+        try:
+            return self.completion_matches[state]
+        except IndexError:
+            return None
+
+
+def _connect(host, port):
+    with closing(socket.create_connection((host, port))) as conn:
+        sockfile = conn.makefile("rwb")
+
+    remote_pdb = _RemotePdb(sockfile)
```

Comment:
We need to think about the conflicts between `_RemotePdb` and `Pdb` - what if the users have both for their process. Also, what if there are two clients trying to connect to the process?

I think to begin with, we can simply reject a separate connection attempt.

The thing I'm concerned here is there's an implicit cache for pdb instance - take a look at `Pdb._last_pdb_instance`. So now if a remote connection is made, `Pdb._last_pdb_instance` will be set to the remote debugger, and the following `breakpoint()`s will use that instance, which will not give what the user wanted.

This might not be a super common case, but we should deal with it. I think clearing `Pdb._last_pdb_instance` after detach would just work.

Reply:
> Separately, I think a hardcoded `breakpoint()` call should work while a remote PDB is attached (that is, it should cause the already attached remote PDB debugger to stop when the `breakpoint()` is hit).

I've tested, it Just Works™.

> > I think clearing `Pdb._last_pdb_instance` after detach would just work.
> 
> Ooh, I thought that was happening already! I agree, that sounds like the right thing to do.

Done in 325f166626840f840cc8c01e26ab6efbb9d6264a

> I think to begin with, we can simply reject a separate connection attempt.

Done in 72830e2ce1bb785727bde0c10de017f81d0566b7